integration_tests/failing_test.py

```diff
@@ -0,0 +1,220 @@
+  ( -1.48219896346314,   2.86114206367024,   2.65169764346032,  0.581191968886918, -0.879586668375104,   7.49819940764743,),
+  (-0.749509673982231,   1.44680552896073,   1.34089490361381,  0.293893744267456, -0.444784224860468,   3.79164547541484,),
+  ( -0.0168203845013521,   0.0324689942512721,   0.0300921637673521,  0.00659551964800653, -0.00998178134584882,    0.085091543182406,),
+  (  0.71586890497956,  -1.38186754045825,  -1.28071057607916, -0.280702704971456,   0.42482066216879,   -3.6214623890502,),
+  (  1.44855819446047,  -2.79620407516777,  -2.59151331592568, -0.568000929590919,  0.859623105683428,  -7.32801632128279,),
+  ( 4.32404297517277,  1.21982063469026, 0.452032773303998, -2.28922571425267, -3.87900483576067, -3.82284004908829,),
+  ( 2.84669419676118, 0.803057772968491, 0.297591647423194, -1.50709083911895, -2.55370740267151, -2.51673182837667,),
+  (  1.36934541834959,  0.386294911246725,  0.143150521542389, -0.724955963985235,  -1.22840996958235,  -1.21062360766505,),
+  ( -0.108003360062012, -0.0304679504750435, -0.0112906043384157,  0.0571789111484879,  0.0968874635068131,  0.0954846130465814,),
+  ( -1.5853521384736, -0.44723081219681, -0.16573173021922, 0.839313786282207,  1.42218489659597,   1.4015928337582,),
+  ( -3.06270091688519, -0.863993673918576, -0.320172856100025,   1.62144866141593,   2.74748232968513,   2.70770105446982,),
+  ( 495.976387330004,  47.2203880990181, -1215.62277241977,  15.2833434534975,   259.00495538446,   35.609827997935,),
+  (   3.09165000587569,   0.206001554376822,   -7.28005326053692, 0.00859894280373505,    1.58972240977151,  0.0665918328475172,),
+  (   1.59254089739601,   0.106113531495721,   -3.75003073782357, 0.00442940438383757,    0.81888245702346,  0.0343021419114977,),
+  (  0.0934317889163547,  0.00622550861462227,    -0.22000821511025, 0.000259865963940125,   0.0480425042754152,  0.00201245097547842,),
+  (   -1.40567731956332,  -0.0936625142664778,      3.3100143076031, -0.00390967245595735,   -0.722797448472637,  -0.0302772399605411,),
+  (   -2.90478642804299,   -0.193550537147578,     6.84003683031646, -0.00807921087585483,    -1.49363740122069,  -0.0625669308965606,),
+  (-39.4174266827821,  49.2517941577222,  6.10806420820106,  5.13048157142721, -14.2523224323315,  12.2507030965516,),
+  ])
+
+from toppra.cpp import PiecewisePolyPath
+path = PiecewisePolyPath.constructHermite(qs, vs, ts)
+
+from example_robot_data.robots_loader import loadUR
+robot = loadUR(limited=True)
+
+constraints = [
+        joint_velocity_constraint(robot),
+        torque_constraint(robot),
+        ]
+constraints_cpp = [
+        joint_velocity_constraint_cpp(robot),
+        torque_constraint_cpp(robot),
+        ]
+
+Ngripoints = len(ts)-1
+
+from toppra.algorithm import TOPPRA
+algo = TOPPRA(constraints, path, solver_wrapper='seidel', gridpoints=
+        np.linspace(0, path.duration, Ngripoints+1))
+sdd, sd, _ = algo.compute_parameterization(0, 0)
```

Comment:
Just a comment, normally the number of grid points should be at least a few times the waypoint, especially if the path is not smooth in-between waypoints.

That said, definitely, the cpp implementation is having a numerical issue because the algorithm should not fail in the forward pass regardless of the number of waypoints. I will help to look into this.

Reply:
I noticed that in the readme, thanks for remembering it. My intent here was to give the same points to TOPPRA as in my case in order to replicate its evaluation.
I saved the values of q, v and a at each gridpoint. To replicate TOPPRA evaluation, I must give a path and gridpoints which will give the same q, v and a.

Reply:
> I will help to look into this.

Thank you for this.

Reply:
The problem is due to the way floats are treated in C++ Apparently 

```
      l[1] = std::max(l[1], box.x[i][0]);
      h[1] = std::min(h[1], box.x[i][1]);
```
Can lead to l[1] > h[1] even if x[i][0] <= x[i][1]. The difference is very small, similar to what you have found ealier. http://www.cplusplus.com/forum/beginner/235970/

I think we need to do what you have done earlier (using cwiseMax). To ensure that the constraints are correctly formed, we will need to put assertions in the preparation steps.

I made the change locally and manage to run your failing test. The velocity differences are ~ 1e-8. 

Note that the python instance needs to use interpolation scheme as well:

```
modified   integration_tests/failing_test.py
@@ -2,13 +2,18 @@ import pinocchio
 import numpy as np
 from toppra.cpp import Interpolation
 
+import toppra
+toppra.setup_logging(level="DEBUG")
+
 def torque_constraint(robot):
-    from toppra.constraint import JointTorqueConstraint
+    from toppra.constraint import JointTorqueConstraint, DiscretizationType
     def inv_dyn (q, v, a):
         return pinocchio.rnea(robot.model, robot.data, q, v, a)
-    return JointTorqueConstraint (inv_dyn,
+    c = JointTorqueConstraint (inv_dyn,
         np.vstack([-robot.model.effortLimit,robot.model.effortLimit]).T,
         np.zeros(robot.model.nv))
+    c.discretization_type = DiscretizationType.Interpolation
+    return c
 
```

Reply:
> The problem is due to the way floats are treated in C++ Apparently
> 
> ```
>       l[1] = std::max(l[1], box.x[i][0]);
>       h[1] = std::min(h[1], box.x[i][1]);
> ```
> 
> Can lead to l[1] > h[1] even if x[i][0] <= x[i][1]. The difference is very small, similar to what you have found ealier. http://www.cplusplus.com/forum/beginner/235970/

I don't think it can be that. Rounding errors in C++ are not different than in Python. Floating point comparison is transitive.

I checked that values in the place above for the current failing test. The value for `l[1]` comes from `x[0]` while `h[1]` comes from `box.x[2][1]`. So definitely, there is a constraint which is in conflicts with the search interval.

Reply:
It's possible that there are multiple box constraints that are conflicting, in which case the condition cwiseMax will create a bug.

Though, each i is treated separately, so box.x[0] wouldn't relate to box.x[2]. Did you have a typo?

Reply:
Ok I think I find out the actual cause, it's because `xNext` conflict with `box.x[i]`. Most likely there is a rounding error in the previous forward step.

Reply:
> Though, each i is treated separately, so box.x[0] wouldn't relate to box.x[2]. Did you have a typo?

No, it is not a typo. `x[0]` does not refer to `box.x[0]` but to the first value of the vector `x` passed as argument. On my computer, `x[0]` and `box.x[2][1]` conflicts. `x[0]` coming from the integration in the forward pass

    m_data.parametrization(i + 1) =
      std::max(0., m_data.parametrization(i) + 2 * deltas(i) * solution(0));

So I agree that the issue is a rounding error in this line.

Reply:
This diff works for me. Without the cwiseMin/Max, the assertions will fire. Look like this is the issue. 

```
@@ -150,6 +158,11 @@ bool qpOASESWrapper::solveStagewiseOptim(std::size_t i,
   if (res == qpOASES::SUCCESSFUL_RETURN) {
     solution.resize(nbVars());
     m_impl->qp.getPrimalSolution(solution.data());
+    TOPPRA_LOG_DEBUG("solution: " << solution);
+    solution = solution.cwiseMax(l.transpose());
+    solution = solution.cwiseMin(h.transpose());
+    assert((solution.transpose().array() <= h.array()).all());
+    assert((solution.transpose().array() >= l.array()).all());
     return true;
   }
```

Reply:
This looks like a good solution to me !

Reply:
I added this fix in #120.